### PR TITLE
Robustness in Order Matching

### PR DESCRIPTION
**Pull request by hysz**
## Description

Changes to Order Matching Implementation:
The algorithm is now written out fully within `calculateMatchedFillResults`. 

Changes to Order Matching Tests:
1. Test author must compute the expected transfer amounts by hand
2. These expected amounts are tested against logs+return values, exchange state, and balances
3. Assertions are more granular to provide better feedback to test authors
4. Additional edge case tests were added

## Checklist:

*   [x] Prefix PR title with `[WIP]` if necessary.
*   [ ] Prefix PR title with bracketed package name(s) corresponding to the changed package(s). For example: `[sol-cov] Fixed bug`.
*   [x] Add tests to cover changes as needed.
*   [ ] Update documentation as needed.
*   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol [hunk 1]
```diff
@@ -162,62 +162,74 @@ contract MixinMatchOrders is
...
-                leftTakerAssetFilledAmount
+        // Calculate fill results for maker and taker assets
+        if (leftTakerAssetAmountRemaining >= rightMakerAssetAmountRemaining) {
+            // Case 1: Right order is fully filled
+            matchedFillResults.right.makerAssetFilledAmount = rightMakerAssetAmountRemaining;
+            matchedFillResults.right.takerAssetFilledAmount = rightTakerAssetAmountRemaining;
+            matchedFillResults.left.makerAssetFilledAmount = getPartialAmount(
+                leftOrder.makerAssetAmount,
+                leftOrder.takerAssetAmount,
+                matchedFillResults.right.makerAssetFilledAmount
```
Comment (abandeali1) [inline packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol]: We can save an mload here by reusing`rightMakerAssetAmountRemaining`

File: packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol [hunk 2]
```diff
@@ -162,62 +162,74 @@ contract MixinMatchOrders is
...
+            // Case 2: Left order is fully filled
+            matchedFillResults.left.makerAssetFilledAmount = leftMakerAssetAmountRemaining;
+            matchedFillResults.left.takerAssetFilledAmount = leftTakerAssetAmountRemaining;
+            matchedFillResults.right.makerAssetFilledAmount = matchedFillResults.left.takerAssetFilledAmount;
+            // @TODO: Use ceil
+            matchedFillResults.right.takerAssetFilledAmount = getPartialAmount(
                 rightOrder.takerAssetAmount,
-                rightTakerAssetFilledAmount
+                rightOrder.makerAssetAmount,
+                matchedFillResults.left.takerAssetFilledAmount
```
Comment (abandeali1) [inline packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol]: Same as above.

File: packages/contracts/test/exchange/match_orders.ts [hunk 3]
```diff
@@ -13,21 +12,18 @@ import { ERC721ProxyContract } from '../../generated_contract_wrappers/erc721_pr
...
-import { ERC20BalancesByOwner, ERC721TokenIdsByOwner, OrderInfo, OrderStatus } from '../utils/types';
+import { ERC20BalancesByOwner, ERC721TokenIdsByOwner } from '../utils/types';
 import { provider, txDefaults, web3Wrapper } from '../utils/web3_wrapper';
 
-chaiSetup.configure();
-const expect = chai.expect;
 const blockchainLifecycle = new BlockchainLifecycle(web3Wrapper);
 
-describe('matchOrders', () => {
+describe.only('matchOrders', () => {
```
Comment (abandeali1) [inline packages/contracts/test/exchange/match_orders.ts]: `only`

File: packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol [hunk 4]
```diff
@@ -162,62 +162,73 @@ contract MixinMatchOrders is
...
+        );
 
-            // The right order receives an amount proportional to how much was spent.
-            rightTakerAssetFilledAmount = getPartialAmountFloor(
-                rightOrder.takerAssetAmount,
-                rightOrder.makerAssetAmount,
-                leftTakerAssetFilledAmount
+        // Calculate fill results for maker and taker assets
+        if (leftTakerAssetAmountRemaining >= rightMakerAssetAmountRemaining) {
+            // Case 1: Right order is fully filled
```
Comment (Recmo) [inline packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol]: Can we add documentation how this follows from `leftTakerAssetAmountRemaining >= rightMakerAssetAmountRemaining`?

File: packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol [hunk 5]
```diff
@@ -162,62 +162,73 @@ contract MixinMatchOrders is
...
-            rightTakerAssetFilledAmount = rightTakerAssetAmountRemaining;
-
-            // The left order receives an amount proportional to how much was spent.
-            leftTakerAssetFilledAmount = getPartialAmountFloor(
-                rightOrder.makerAssetAmount,
+            // Case 2: Left order is fully filled
+            matchedFillResults.left.makerAssetFilledAmount = leftMakerAssetAmountRemaining;
+            matchedFillResults.left.takerAssetFilledAmount = leftTakerAssetAmountRemaining;
+            matchedFillResults.right.makerAssetFilledAmount = matchedFillResults.left.takerAssetFilledAmount;
+            matchedFillResults.right.takerAssetFilledAmount = getPartialAmountCeil(
```
Comment (Recmo) [inline packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol]: Add a note why we need to round up here.

File: packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol [hunk 6]
```diff
@@ -162,62 +162,73 @@ contract MixinMatchOrders is
...
             );
+            matchedFillResults.left.takerAssetFilledAmount = matchedFillResults.right.makerAssetFilledAmount;
         } else {
-            // Right order will be fully filled: maximally fill right
-            rightTakerAssetFilledAmount = rightTakerAssetAmountRemaining;
-
-            // The left order receives an amount proportional to how much was spent.
-            leftTakerAssetFilledAmount = getPartialAmountFloor(
-                rightOrder.makerAssetAmount,
+            // Case 2: Left order is fully filled
```
Comment (Recmo) [inline packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol]: Add a note how this follows from `leftTakerAssetAmountRemaining < rightMakerAssetAmountRemaining`

File: packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol [hunk 7]
```diff
@@ -162,62 +162,73 @@ contract MixinMatchOrders is
...
+            matchedFillResults.left.makerAssetFilledAmount = leftMakerAssetAmountRemaining;
+            matchedFillResults.left.takerAssetFilledAmount = leftTakerAssetAmountRemaining;
+            matchedFillResults.right.makerAssetFilledAmount = matchedFillResults.left.takerAssetFilledAmount;
```
Comment (Recmo) [inline packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol]: Maybe use `leftTakerAssetAmountRemaining` here and below instead of `matchedFillResults.left.takerAssetFilledAmount`? It's a bit simpler and we are not reading what we just wrote.

File: packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol [hunk 8]
```diff
@@ -162,62 +162,73 @@ contract MixinMatchOrders is
...
-            rightTakerAssetFilledAmount = getPartialAmountFloor(
-                rightOrder.takerAssetAmount,
-                rightOrder.makerAssetAmount,
-                leftTakerAssetFilledAmount
+        // Calculate fill results for maker and taker assets
+        if (leftTakerAssetAmountRemaining >= rightMakerAssetAmountRemaining) {
+            // Case 1: Right order is fully filled
+            matchedFillResults.right.makerAssetFilledAmount = rightMakerAssetAmountRemaining;
+            matchedFillResults.right.takerAssetFilledAmount = rightTakerAssetAmountRemaining;
+            matchedFillResults.left.makerAssetFilledAmount = getPartialAmountFloor(
```
Comment (abandeali1) [inline packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol]: Huge nit, but this is would be a bit more readable if it was:

```
            matchedFillResults.left.takerAssetFilledAmount = matchedFillResults.right.makerAssetFilledAmount;
            matchedFillResults.left.makerAssetFilledAmount = getPartialAmountFloor(
                leftOrder.makerAssetAmount,
                leftOrder.takerAssetAmount,
                matchedFillResults.left.takerAssetFilledAmount
            );
```

File: packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol [hunk 9]
```diff
@@ -162,62 +162,73 @@ contract MixinMatchOrders is
...
-                rightOrder.makerAssetAmount,
+            // Case 2: Left order is fully filled
+            matchedFillResults.left.makerAssetFilledAmount = leftMakerAssetAmountRemaining;
+            matchedFillResults.left.takerAssetFilledAmount = leftTakerAssetAmountRemaining;
+            matchedFillResults.right.makerAssetFilledAmount = matchedFillResults.left.takerAssetFilledAmount;
+            matchedFillResults.right.takerAssetFilledAmount = getPartialAmountCeil(
                 rightOrder.takerAssetAmount,
-                rightTakerAssetFilledAmount
+                rightOrder.makerAssetAmount,
+                matchedFillResults.left.takerAssetFilledAmount
```
Comment (abandeali1) [inline packages/contracts/src/2.0.0/protocol/Exchange/MixinMatchOrders.sol]: Again, a bit easier to understand as `matchedFillResults.right.makerAssetFilledAmount` instead of `matchedFillResults.left.takerAssetFilledAmount`
